api/demux.go

```diff
@@ -0,0 +1,46 @@
+package api
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+// DemuxHeader reads header for stream from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxHeader(r io.Reader, buffer []byte) (fd, sz int, err error) {
+	n, err := io.ReadFull(r, buffer[0:8])
+	if err != nil {
+		return
+	}
+	if n < 8 {
+		err = io.ErrUnexpectedEOF
+		return
+	}
+
+	fd = int(buffer[0])
+	if fd < 0 || fd > 3 {
+		err = fmt.Errorf(`channel "%d" found, 0-3 supported`, fd)
+		return
+	}
+
+	sz = int(binary.BigEndian.Uint32(buffer[4:8]))
+	return
+}
+
+// DemuxFrame reads contents for frame from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxFrame(r io.Reader, buffer []byte, length int) (frame []byte, err error) {
+	if len(buffer) < length {
+		buffer = append(buffer, make([]byte, length-len(buffer)+1)...)
+	}
+
+	n, err := io.ReadFull(r, buffer[0:length])
+	if err != nil {
+		return nil, nil
```

Comment:
Hmm. This is a shameless copy of DemuxFrame() from  https://github.com/containers/podman/blob/main/pkg/bindings/containers/attach.go but you're probably right. Should i change it?

Reply:
Oh interesting. I think this may cause a `panic` in Podman then. If there's an error in `DemuxFrame` they will try to access a `nil` slice in the `switch`: https://github.com/containers/podman/blob/main/pkg/bindings/containers/attach.go#L230.

I think we should change it. Returning `nil, nil` doesn't seem like the right thing. They also have a similar code on `DemuxHeader` that [just returns](https://github.com/containers/podman/blob/main/pkg/bindings/containers/attach.go#L257-L260), which is what I would expect to happen here as well.

Reply:
IC. I will improve it. Best to create a podman issue as well, will do so.

Reply:
Yeah, that's a good point. Maybe I'm missing something here and they did this for a reason. Let me know if you get a reply because I'm curious about it as well 😄